### DISCUSSION TITLE
Use of Custom TMS (Non-Earth)

### DISCUSSION THREAD

**Discussion by jlaura** [developmentseed/titiler#427] [Q&A]
I am seeing the following error attempting to use a custom TMS (for non-earth data). I wonder if anything jumps out at anyone? I have a working example just using riotiler (many thanks @vincentsarago for seeing that issue and posting the example!). The next step, is to get a titiler example working. Then I can deploy body specific tilers to AWS (that's the plan anyway!).

Here is the error I am seeing:

```
rasterio._err.CPLE_NotSupportedError: Cannot find coordinate operations from 'EPSG:3857' to 'PROJCRS["Equirectangular EUROPA",BASEGEOGCRS["GCS_EUROPA",DATUM["D_EUROPA",ELLIPSOID["EUROPA_localRadius",1560800,0,LENGTHUNIT["metre",1,ID["EPSG",9001]]]],PRIMEM["Reference_Meridian",0,ANGLEUNIT["degree",0.0174532925199433,ID["EPSG",9122]]]],CONVERSION["unnamed",METHOD["Equidistant Cylindrical (Spherical)",ID["EPSG",1029]],PARAMETER["Latitude of 1st standard parallel",0,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8823]],PARAMETER["Longitude of natural origin",180,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8802]],PARAMETER["False easting",0,LENGTHUNIT["metre",1],ID["EPSG",8806]],PARAMETER["False northing",0,LENGTHUNIT["metre",1],ID["EPSG",8807]]],CS[Cartesian,2],AXIS["easting",east,ORDER[1],LENGTHUNIT["metre",1,ID["EPSG",9001]]],AXIS["northing",north,ORDER[2],LENGTHUNIT["metre",1,ID["EPSG",9001]]]]'
```

So, no transformation operations from Europa to 4326. Since both are just lat/lon, I was anticipating that this hack would work? Perhaps I am messing up the configuration of the custom WMS? 

Here is a minimal reproducible app:

dependencies.py
```
from enum import Enum
from morecantile import tms, TileMatrixSet
from pyproj import CRS
from fastapi import Query

# 1. Create Custom TMS
europa_crs = CRS.from_authority("ESRI", 104915)
europa_tms = TileMatrixSet.custom(
    crs=europa_crs, extent=europa_crs.area_of_use.bounds, matrix_scale=[2, 1],
    name='Europa'
)

# 2. Register TMS
tms = tms.register([europa_tms,])
print(tms)
# 3. Create ENUM with available TMS
TileMatrixSetNames = Enum(  # type: ignore
    "TileMatrixSetNames", [(a, a) for a in sorted(tms.list())]
)

print('SetNames', TileMatrixSetNames)

# 4. Create Custom TMS dependency
def TMSParams(
    TileMatrixSetId: TileMatrixSetNames = Query(
        TileMatrixSetNames.WebMercatorQuad,  # type: ignore
        description="TileMatrixSet Name (default: 'WebMercatorQuad')",
    )
) -> TileMatrixSet:
    """TileMatrixSet Dependency."""
    return tms.get(TileMatrixSetId.name)
```

routes.py
```
from titiler.core.factory import TilerFactory, TMSFactory
from .dependencies import TileMatrixSetNames, TMSParams

tms = TMSFactory(supported_tms=TileMatrixSetNames, tms_dependency=TMSParams)
cog = TilerFactory(tms_dependency=TMSParams)
```

main.py
```
from titiler.core.errors import DEFAULT_STATUS_CODES, add_exception_handlers

from fastapi import FastAPI

from .routes import cog, tms

app = FastAPI(title="My simple app with custom TMS")

app.include_router(cog.router, tags=["Cloud Optimized GeoTIFF"])
app.include_router(tms.router, tags=["TileMatrixSets"])
add_exception_handlers(app, DEFAULT_STATUS_CODES)
```

web map (Jupyter)
```
bounds = (90, 7.8, 90.5, 8.3)

m = Map(
    center=(
        (bounds[1] + bounds[3]) / 2,
        (bounds[0] + bounds[2]) / 2
    ),
    zoom=4,
    crs=projections.EPSG4326,  # HACK: the europa TMS is in degree and covers -180, -90, 180, 90 like the WGS84
)

layer = TileLayer(
    url="http://127.0.0.1:8000/cog/tiles/{z}/{x}/{y}?url=https://asc-jupiter.s3.us-west-2.amazonaws.com/europa/galileo_voyager/controlled_images/equi/s0383713765/s0383713765_photr.tif",
    min_zoom=4,
    max_zoom=6,
    opacity=1,
)
m.add_layer(layer)
m
```

**Comment by vincentsarago** [comment]
@jlaura 👁️ your code it seems that the default TMS used is WebMercator, can you change to `Europa` 

```python
# 4. Create Custom TMS dependency
def TMSParams(
    TileMatrixSetId: TileMatrixSetNames = Query(
        TileMatrixSetNames. Europa,  # type: ignore
        description="TileMatrixSet Name (default: 'Europa')",
    )
) -> TileMatrixSet:
    """TileMatrixSet Dependency."""
    return tms.get(TileMatrixSetId.name)
```

**Reply by jlaura** [reply]
Thanks for the super fast answer. I think I am still confused as I am still seeing the coordinate conversion error:

```
rasterio._err.CPLE_NotSupportedError: Cannot find coordinate operations from 'EPSG:3857' to 'PROJCRS["Equirectangular EUROPA",BASEGEOGCRS["GCS_EUROPA",DATUM["D_EUROPA",ELLIPSOID["EUROPA_localRadius",1560800,0,LENGTHUNIT["metre",1,ID["EPSG",9001]]]],PRIMEM["Reference_Meridian",0,ANGLEUNIT["degree",0.0174532925199433,ID["EPSG",9122]]]],CONVERSION["unnamed",METHOD["Equidistant Cylindrical (Spherical)",ID["EPSG",1029]],PARAMETER["Latitude of 1st standard parallel",0,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8823]],PARAMETER["Longitude of natural origin",180,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8802]],PARAMETER["False easting",0,LENGTHUNIT["metre",1],ID["EPSG",8806]],PARAMETER["False northing",0,LENGTHUNIT["metre",1],ID["EPSG",8807]]],CS[Cartesian,2],AXIS["easting",east,ORDER[1],LENGTHUNIT["metre",1,ID["EPSG",9001]]],AXIS["northing",north,ORDER[2],LENGTHUNIT["metre",1,ID["EPSG",9001]]]]'
```

After the changes that you made to [riotiler](https://cogeotiff.github.io/rio-tiler/examples/Using-nonEarth-dataset/) (thank you!) I have followed the example and get a local tile server. I created an env with riotiler 3.0.3 and titiler.application 0.4.0 in an attempt to get access to all of the API endpoints, etc. I created a [minimal app](https://devseed.com/titiler/intro/#customized-minimal-app), but see the above (same error for transformation to 4326).

I thought that I needed to create a custom TMS, but perhaps not? Should I be looking someplace else to take the [example here](https://cogeotiff.github.io/rio-tiler/examples/Using-nonEarth-dataset/#tile-server) and get it integrated into a custom titiler deploy? The line `with COGReader(self.url, tms=europa_tms) as cog:` seems to be the place where the COG is read with the proper TMS.

Thanks!

**Reply by jlaura** [reply]
I just found these docs which look promising for what I am trying to accomplish!

https://devseed.com/titiler/advanced/tiler_factories/

**Reply by vincentsarago** [reply]
I'm not sure to follow everything.

Basically, the error your are seeing is because rasterio is trying to transform the tile bounds (in web mercator) to the dataset coordinates system (`PROJCS["Equirectangular EUROPA",GEOGCS["GCS_EUROPA",DATUM["D_EUROPA",SPHEROID["EUROPA_localRadius",1560800,0]],PRIMEM["Reference_Meridian",0],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]]],PROJECTION["Equirectangular"],PARAMETER["standard_parallel_1",0],PARAMETER["central_meridian",180],PARAMETER["false_easting",0],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH]]`). So the first thing you need to make sure is to use a TMS that is compatible with your dataset, which is what you did by creating a custom TMS (europa_tms).

your app should be like 👇 
```python
from enum import Enum
from morecantile import tms, TileMatrixSet
from pyproj import CRS
from fastapi import Query

from titiler.core.factory import TilerFactory, TMSFactory

from fastapi import FastAPI

# 1. Create Custom TMS
europa_crs = CRS.from_authority("ESRI", 104915)
europa_tms = TileMatrixSet.custom(
    crs=europa_crs, extent=europa_crs.area_of_use.bounds, matrix_scale=[2, 1],  # type: ignore
    identifier='Europa'
)

# 2. Register TMS
tms = tms.register([europa_tms])

# 3. Create ENUM with available TMS (fixed to only europa_tms)
TileMatrixSetNames = Enum(  # type: ignore
    "TileMatrixSetNames", [("Europa", "Europa")]
)


# 4. Create Custom TMS dependency
def TMSParams(
    TileMatrixSetId: TileMatrixSetNames = Query(
        TileMatrixSetNames.Europa,  # type: ignore
        description="TileMatrixSet Name (default: 'Europa')",
    )
) -> TileMatrixSet:
    """TileMatrixSet Dependency."""
    return tms.get(TileMatrixSetId.name)


tms_endpoints = TMSFactory(supported_tms=TileMatrixSetNames, tms_dependency=TMSParams)
cog_endpoints = TilerFactory(tms_dependency=TMSParams)


app = FastAPI(title="My simple app with custom TMS")
app.include_router(cog_endpoints.router, tags=["Cloud Optimized GeoTIFF"])
app.include_router(tms_endpoints.router, tags=["TileMatrixSets"])
```

Sadly this doesn't seem to work and I'm now getting 

```
asterio._err.CPLE_NotSupportedError: Cannot find coordinate operations from 'ESRI:104915' to 'PROJCRS["Equirectangular EUROPA",BASEGEOGCRS["GCS_EUROPA",DATUM["D_EUROPA",ELLIPSOID["EUROPA_localRadius",1560800,0,LENGTHUNIT["metre",1,ID["EPSG",9001]]]],PRIMEM["Reference_Meridian",0,ANGLEUNIT["degree",0.0174532925199433,ID["EPSG",9122]]]],CONVERSION["unnamed",METHOD["Equidistant Cylindrical (Spherical)",ID["EPSG",1029]],PARAMETER["Latitude of 1st standard parallel",0,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8823]],PARAMETER["Longitude of natural origin",180,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8802]],PARAMETER["False easting",0,LENGTHUNIT["metre",1],ID["EPSG",8806]],PARAMETER["False northing",0,LENGTHUNIT["metre",1],ID["EPSG",8807]]],CS[Cartesian,2],AXIS["easting",east,ORDER[1],LENGTHUNIT["metre",1,ID["EPSG",9001]]],AXIS["northing",north,ORDER[2],LENGTHUNIT["metre",1,ID["EPSG",9001]]]]'
```

Which is either an error coming from my gdal/proj install or in the custom Europa TMS you provided

**Reply by jlaura** [reply]
@vincentsarago Cool! We are 100% on the same page. Here is a working riotiler example:

## Working riotiler example with the TMS

server.py
```
import json
from concurrent import futures

from tornado import web
import tornado.ioloop
from tornado import gen
from tornado.httpserver import HTTPServer
from tornado.concurrent import run_on_executor

from rio_tiler.io import COGReader
from rio_tiler.errors import TileOutsideBounds
from rio_tiler.profiles import img_profiles

from pyproj import CRS
from morecantile import TileMatrixSet

# Create a CUSTOM TMS using the europa ESRI:104915 projection
europa_crs = CRS.from_authority("ESRI", 104915)
europa_tms = TileMatrixSet.custom(
    crs=europa_crs, extent=europa_crs.area_of_use.bounds, matrix_scale=[2, 1],
)

class TileServer:
    def __init__(self, src_path):
        """Initialize Tornado app."""
        self.server = None
        self.app = web.Application([
            (r"^/tiles/(\d+)/(\d+)/(\d+)", TileHandler, {"url": src_path}),
        ])

    def start(self):
        """Start tile server."""
        self.server = HTTPServer(self.app)
        self.server.listen(8080)
  
    def stop(self):
        """Stop tile server."""
        if self.server:
            self.server.stop()


class TileHandler(web.RequestHandler):
    """Tile requests handler."""

    executor = futures.ThreadPoolExecutor(max_workers=16)

    def initialize(self, url):
        """Initialize tiles handler."""
        self.url = url

    @run_on_executor
    def _get_tile(self, z, x, y):

        try:
            with COGReader(self.url, tms=europa_tms) as cog:
                data = cog.tile(x, y, z)
        except TileOutsideBounds:
            raise web.HTTPError(404)

        image = data.post_process(in_range=((0, .5),))

        prof = img_profiles.get("PNG", {})
        return image.render(img_format="PNG", **prof)

    @gen.coroutine
    def get(self, z, x, y):
        """Retunrs tile data and header."""
        self.set_header("Access-Control-Allow-Origin", "*")
        self.set_header("Access-Control-Allow-Methods", "GET")
        self.set_header("Content-Type", "image/png")
        self.set_header("Cache-Control", "no-store, no-cache, must-revalidate")
        res = yield self._get_tile(int(z), int(x), int(y))
        self.write(res)

src_path = 'https://asc-jupiter.s3.us-west-2.amazonaws.com/europa/galileo_voyager/controlled_images/equi/s0383713765/s0383713765_photr.tif'
ts = TileServer(src_path)
ts.start()
tornado.ioloop.IOLoop.instance().start()
```

map (in a notebook)
```
bounds = (90, 7.8, 90.5, 8.3)

m = Map(
    center=(
        (bounds[1] + bounds[3]) / 2,
        (bounds[0] + bounds[2]) / 2
    ),
    zoom=4,
    crs=projections.EPSG4326,  # HACK: the europa TMS is in degree and covers -180, -90, 180, 90 like the WGS84
)

layer = TileLayer(
    url="http://127.0.0.1:8080/tiles/{z}/{x}/{y}",
    min_zoom=4,
    max_zoom=6,
    opacity=1,
)
m.add_layer(layer)
m
```

## Output Example
<img width="1064" alt="Screen Shot 2022-01-25 at 2 16 33 PM" src="https://user-images.githubusercontent.com/2105468/151061166-11ddebaa-8cb2-4ed8-80fc-347df3181204.png">

## More information
- I have the riotiler example in a different environment than the titiler example. Comparing the environments the GDAL and Proj versions are identical.
- Should I be seeing the custom TMS in the /tileMatrixSets endpoint? I wonder if the dependency is injecting correctly as I do not see it.

Many thanks!

**Reply by vincentsarago** [reply]
Oh I see, so it works in rio-tiler but not in TiTiler 👌 

> I have the riotiler example in a different environment than the titiler example. Comparing the environments the GDAL and Proj versions are identical.

👌 

> Should I be seeing the custom TMS in the /tileMatrixSets endpoint? I wonder if the dependency is injecting correctly as I do not see it.

yes it works, the error message I'm seeing exactly tells that it tries to do a transformation between Europa TMS's CRS (`CRS.from_authority("ESRI", 104915)`) and the dataset internal CRS

I have absolutely no idea why titiler and rio-tiler behave differently 🤷

**Reply by vincentsarago** [reply]
well the rio-tiler example doesn't work for me 🤷‍♂️ 

it's weird because even the rio-tiler notebook doesn't work for me anymore :-(

**Reply by vincentsarago** [reply]
Oh I found why it was failing on my side. You need to make sure that rasterio/gdal use PROJ>=8.0! (FYI, homebrew compile gdal with proj7)

**Reply by jlaura** [reply]
Wonderful and thank you for trying to reproduce! 

I have proj 8.2.1, libgdal 3.4.1, Rasterio 1.2.10. I am working on learning the code base. It seems to me that ti-tiler is using riotiler to do the read through the [Tiler Factory](https://developmentseed.org/titiler/advanced/tiler_factories/). Could it be the case that the injection of the custom TMS is not being caught by the reader?  

I think the error means that this is not the case, but I'm not 100% sure where the issue could be otherwise.

**Reply by vincentsarago** [reply]
>  Could it be the case that the injection of the custom TMS is not being caught by the reader?

I don't think, it's well tested!

**Reply by jlaura** [reply]
I have been able to get Mars and Europa support working! Whatever the issue was, the new releases have fixed it.By modifying the `custom.py` file I can set:

```python
mars_crs = CRS.from_authority("ESRI", 104971)
ESRI104971 = morecantile.TileMatrixSet.custom(
    crs=mars_crs, 
    identifier="Mars",
    extent=mars_crs.area_of_use.bounds, 
    matrix_scale=[2, 2],
)

# CUSTOM TMS for Europa
europa_crs = CRS.from_authority("ESRI", 104915)
ESRI104915 = morecantile.TileMatrixSet.custom(
    crs=europa_crs, 
    identifier="Europa",
    extent=europa_crs.area_of_use.bounds, 
    matrix_scale=[2, 2],
)

tms = tms.register([ESRI104915, ESRI104971])
```

and then 

```python
def TMSParams(
    TileMatrixSetId: TileMatrixSetName = Query(
        TileMatrixSetName.Europa,  # type: ignore
        description="TileMatrixSet Name (default: 'Europa')",
    )
) -> TileMatrixSet:
    """TileMatrixSet Dependency."""
    return tms.get(TileMatrixSetId.name)
```

I then can use the `/cog/tiles` endpoint to get Europa tiles.

Two questions remain:

1. Is it possible to deploy with multiple different TMSParams being passed to the TileFactory? This way a single service could support multiple different bodies? Or, should I deploy multiple lambdas, one for each body?
2. The data (in S3) are 32-bit. I am having to pass the `rescale=x,y` parameter along with the query string to the URL. Is rio-tiler pulling any kind of no data value information or statistics off of the image before tiling? I am trying to determine if I need to do something different with my data or if I should be needing to pass the rescale bounds.

Thanks!

`titiler.__version__ == 0.4.2` 😉

**Reply by vincentsarago** [reply]
> Is it possible to deploy with multiple different TMSParams being passed to the TileFactory? This way a single service could support multiple different bodies? Or, should I deploy multiple lambdas, one for each body?


(I'm not sure to fully understand what you want here, so I'll take a guess) you can create endpoints using different TMSParams as 
```python

TMSEuropa = Enum(  # type: ignore
    "TMSEuropa", [("Europa", "Europa")]
)

def TMSParamsEurope(
    TileMatrixSetId: TMSEuropa = Query(
        TMSEuropa.Europa,  # type: ignore
        description="TileMatrixSet Name (default: 'Europa')",
    )
) -> TileMatrixSet:
    """TileMatrixSet Dependency."""
    return tms.get(TileMatrixSetId.name)


europa = TilerFactory(tms_dependency= TMSParamsEurope, router_prefix="/europa")

TMSMars = Enum(  # type: ignore
    "TMSMars", [("Mars", "Mars")]
)

def TMSParamsMars(
    TileMatrixSetId: TMSMars= Query(
        TMSMars.Mars,  # type: ignore
        description="TileMatrixSet Name (default: 'Mars')",
    )
) -> TileMatrixSet:
    """TileMatrixSet Dependency."""
    return tms.get(TileMatrixSetId.name)


mars = TilerFactory(tms_dependency=TMSParamsMars, router_prefix="/mars")


app = FastAPI()
app.include_router(europa.router, prefix="/europa")
app.include_router(mars.router, prefix="/mars")
```

☝️ then you'll have 2 set of endpoints `/europa/tiles/...` `/mars/tiles/...` 


> The data (in S3) are 32-bit. I am having to pass the rescale=x,y parameter along with the query string to the URL. Is rio-tiler pulling any kind of no data value information or statistics off of the image before tiling? I am trying to determine if I need to do something different with my data or if I should be needing to pass the rescale bounds.

This is a tricky part with rio-tiler, we don't do any rescaling if the user don't ask to. If the data require rescaling we usually ask the user to first use the `/statistics` endpoint to get the `min/max` and then forward the rescaling parameter to the tile endpoint (or tilejson endpoint) 

FYI: I'm planning to use geotiff metadata to allow automatic rescaling https://github.com/cogeotiff/rio-tiler/issues/488 but I don't have timeline for this
